content/docs/iac/using-pulumi/extending-pulumi/build-a-provider.md

```diff
@@ -0,0 +1,722 @@
+			Content: req.Inputs.Content,
+		},
+	}, nil
+}
+```
+
+##### The `Diff` operation
+
+The `Diff` operation compares a resource's recorded state (if any) to the current input values for the resource, to determine what kind of changes need to be made. The `diff` object is a map of property names to the kind of diff operation (if any) that a property needs to have made. The logic of this function is straightforward: compare `req.Inputs.<propertyName>` to `req.State.<propertyName>` and if they aren't equal, add to the diff that this property needs to be updated using `p.PropertyDiff{Kind: p.Update}`. Only include the properties that need to change in the `diff` map. Finally return a `DiffResponse` containing the `diff` map, and set a few other options to configure the update behavior.
+
+```go
+func (File) Diff(ctx context.Context, req infer.DiffRequest[FileArgs, FileState]) (infer.DiffResponse, error) {
+	diff := map[string]p.PropertyDiff{}
+	if req.Inputs.Content != req.State.Content {
+		diff["content"] = p.PropertyDiff{Kind: p.Update}
+	}
+	if req.Inputs.Force != req.State.Force {
+		diff["force"] = p.PropertyDiff{Kind: p.Update}
+	}
+	if req.Inputs.Path != req.State.Path {
+		diff["path"] = p.PropertyDiff{Kind: p.UpdateReplace}
+	} else {
+		_, err := os.Stat(req.Inputs.Path)
+		if os.IsNotExist(err) {
+			diff["path"] = p.PropertyDiff{Kind: p.Add}
+		}
+	}
+	return infer.DiffResponse{
+		DeleteBeforeReplace: true,
+		HasChanges:          len(diff) > 0,
+		DetailedDiff:        diff,
+	}, nil
+}
+```
+
+#### The `Read` operation
+
+The `Read` operation fetches the resource. The `ReadRequest` has a `ID` property that can be used, in this case, to determine the path to the file, and the base library functions can be used to read the file from disk, populating the `Content` field.
+
+```go
+func (File) Read(ctx context.Context, req infer.ReadRequest[FileArgs, FileState]) (infer.ReadResponse[FileArgs, FileState], error) {
+	path := req.ID
+	byteContent, err := os.ReadFile(path)
+	if err != nil {
+		return infer.ReadResponse[FileArgs, FileState]{}, err
+	}
+	content := string(byteContent)
+	return infer.ReadResponse[FileArgs, FileState]{
+		ID: path,
+		Inputs: FileArgs{
+			Path:    path,
+			Force:   req.Inputs.Force && req.State.Force,
+			Content: content,
+		},
+		State: FileState{
+			Path:    path,
+			Force:   req.Inputs.Force && req.State.Force,
+			Content: content,
+		},
+	}, nil
+}
+```
+
+#### Managing resource output fields
+
+Finally, `WireDependencies` defines the outputs that are made available on the resource, logically connecting the inputs and stored state values.
+
+```go
+func (File) WireDependencies(f infer.FieldSelector, args *FileArgs, state *FileState) {
+	f.OutputField(&state.Content).DependsOn(f.InputField(&args.Content))
+	f.OutputField(&state.Force).DependsOn(f.InputField(&args.Force))
+	f.OutputField(&state.Path).DependsOn(f.InputField(&args.Path))
+}
+```
+
+<!-- TODO: write this section
+
+### Testing with the SDK
```

Comment:
Here's a good example of how one would test the resource:
https://github.com/pulumi/pulumi-go-provider/blob/077df8bd52cec2f24603dc867028ac4fcb1ceaa8/examples/random-login/main_test.go#L169

Reply:
I'm going to come back to this and fill in that section in a separate PR.